### PR TITLE
PR: Restore race between Rope and Jedi for completions

### LINKED ISSUE #2: Restore Jedi/Rope race
@steff456, please see https://github.com/palantir/python-language-server/pull/155 for the initial @andfoy implementation. We need to restore that here.

### PR DESCRIPTION
**Pull request by steff456**
Restores the race between Jedi and Rope for completion.

Fixes #2

### CODE REVIEW

File: pyls/plugins/jedi_completion.py [hunk 1]
```diff
@@ -6,16 +6,19 @@
...
     definitions = document.jedi_script(position).completions()
-    return [{
+    comp = [{
         'label': _label(d),
         'kind': _kind(d),
         'detail': _detail(d),
         'documentation': _utils.format_docstring(d.docstring()),
         'sortText': _sort_text(d),
         'insertText': d.name
-    } for d in definitions] or None
```
Comment (ccordoba12) [inline pyls/plugins/jedi_completion.py]: Why did you remove `None` from here? I think we should leave it.

File: pyls/plugins/rope_completion.py [hunk 2]
```diff
@@ -11,27 +11,30 @@
 @hookimpl
 def pyls_settings():
     # Default rope_completion to disabled
-    return {'plugins': {'rope_completion': {'enabled': False}}}
+    return {'plugins': {'rope_completion': {'enabled': True}}}
```
Comment (ccordoba12) [inline pyls/plugins/rope_completion.py]: Please revert this and leave it as `False` for now.

File: pyls/plugins/rope_completion.py [hunk 3]
```diff
@@ -11,27 +11,30 @@
...
 
 
 @hookimpl
-def pyls_completions(config, workspace, document, position):
+def pyls_completions(document, position, config=None, workspace=None):
```
Comment (ccordoba12) [inline pyls/plugins/rope_completion.py]: Please revert this change and leave the signature as it was before.

File: pyls/plugins/rope_completion.py [hunk 4]
```diff
@@ -11,27 +11,30 @@
...
     word = document.word_at_position({
         # The -1 should really be trying to look at the previous word, but that might be quite expensive
         # So we only skip import completions when the cursor is one space after `import`
-        'line': position['line'], 'character': max(position['character'] - 1, 0),
+        'line': position['line'],
+        'character': max(position['character'] - 1, 0),
     })
     if word == 'import':
-        return None
+        return []
```
Comment (ccordoba12) [inline pyls/plugins/rope_completion.py]: Why this change? I think we should preserve the current functionality.

File: pyls/plugins/rope_completion.py [hunk 5]
```diff
@@ -11,27 +11,30 @@
...
 
     offset = document.offset_at_position(position)
     rope_config = config.settings(document_path=document.path).get('rope', {})
     rope_project = workspace._rope_project_builder(rope_config)
     document_rope = document._rope_resource(rope_config)
 
     try:
-        definitions = code_assist(rope_project, document.source, offset, document_rope, maxfixes=3)
+        definitions = code_assist(rope_project, document.source, offset,
+                                  document_rope, maxfixes=3)
```
Comment (ccordoba12) [inline pyls/plugins/rope_completion.py]: Please revert this. I think we should not try to impose the 79 column width limit we have in Spyder in the PyLS code.

File: pyls/plugins/rope_completion.py [hunk 6]
```diff
@@ -51,8 +54,10 @@ def pyls_completions(config, workspace, document, position):
             'sortText': _sort_text(d)
         })
     definitions = new_definitions
-
-    return definitions or None
```
Comment (ccordoba12) [inline pyls/plugins/rope_completion.py]: Again, I think we should leave this as it and return `None` in case no completions are found, unless you have a very good reason to change it to an empty list.

File: pyls/python_ls.py [hunk 7]
```diff
@@ -147,14 +152,18 @@ def capabilities(self):
         log.info('Server capabilities: %s', server_capabilities)
         return server_capabilities
 
-    def m_initialize(self, processId=None, rootUri=None, rootPath=None, initializationOptions=None, **_kwargs):
-        log.debug('Language server initialized with %s %s %s %s', processId, rootUri, rootPath, initializationOptions)
+    def m_initialize(self, processId=None, rootUri=None, rootPath=None,
+                     initializationOptions=None, **_kwargs):
+        log.debug('Language server initialized with %s %s %s %s', processId,
+                  rootUri, rootPath, initializationOptions)
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Please revert this change. It's purely stylistic

File: pyls/python_ls.py [hunk 8]
```diff
@@ -147,14 +152,18 @@ def capabilities(self):
...
         if rootUri is None:
             rootUri = uris.from_fs_path(rootPath) if rootPath is not None else ''
 
         self.workspace = Workspace(rootUri, self._endpoint)
-        self.config = config.Config(rootUri, initializationOptions or {}, processId)
+        self.config = config.Config(rootUri, initializationOptions or {},
+                                    processId)
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 9]
```diff
@@ -164,9 +173,11 @@ def watch_parent_process(pid):
                     log.info("parent process %s is not alive", pid)
                     self.m_exit()
                 log.debug("parent process %s is still alive", pid)
-                threading.Timer(PARENT_PROCESS_WATCH_INTERVAL, watch_parent_process, args=[pid]).start()
+                threading.Timer(PARENT_PROCESS_WATCH_INTERVAL,
+                                watch_parent_process, args=[pid]).start()
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 10]
```diff
@@ -164,9 +173,11 @@ def watch_parent_process(pid):
...
 
-            watching_thread = threading.Thread(target=watch_parent_process, args=(processId,))
+            watching_thread = threading.Thread(target=watch_parent_process,
+                                               args=(processId,))
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 11]
```diff
@@ -177,26 +188,36 @@ def m_initialized(self, **_kwargs):
         pass
 
     def code_actions(self, doc_uri, range, context):
-        return flatten(self._hook('pyls_code_actions', doc_uri, range=range, context=context))
+        return flatten(self._hook('pyls_code_actions', doc_uri, range=range,
+                                  context=context))
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 12]
```diff
@@ -177,26 +188,36 @@ def m_initialized(self, **_kwargs):
...
 
     def code_lens(self, doc_uri):
         return flatten(self._hook('pyls_code_lens', doc_uri))
 
     def completions(self, doc_uri, position):
-        completions = self._hook('pyls_completions', doc_uri, position=position)
+        completions = _utils.race_hooks(
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Let's make the race optional, i.e. if `rope_completion = True` then we'll do the race but if not we'll only use Jedi.

I think that way we'll have much better chances to get this approved in the PyLS main repo.

File: pyls/python_ls.py [hunk 13]
```diff
@@ -177,26 +188,36 @@ def m_initialized(self, **_kwargs):
...
+            self._hook_caller('pyls_completions'),
+            self._pool,
+            document=self.workspace.get_document(doc_uri) if doc_uri else None,
+            position=position,
+            config=self.config,
+            workspace=self.workspace
+        )
         return {
             'isIncomplete': False,
-            'items': flatten(completions)
```
Comment (ccordoba12) [inline pyls/python_ls.py:189]: Why is this change is necessary?

File: pyls/python_ls.py [hunk 14]
```diff
@@ -177,26 +188,36 @@ def m_initialized(self, **_kwargs):
...
+            'items': completions
         }
 
     def definitions(self, doc_uri, position):
-        return flatten(self._hook('pyls_definitions', doc_uri, position=position))
+        return flatten(self._hook('pyls_definitions', doc_uri,
+                                  position=position))
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 15]
```diff
@@ -177,26 +188,36 @@ def m_initialized(self, **_kwargs):
...
 
     def document_symbols(self, doc_uri):
         return flatten(self._hook('pyls_document_symbols', doc_uri))
 
     def execute_command(self, command, arguments):
-        return self._hook('pyls_execute_command', command=command, arguments=arguments)
+        return self._hook('pyls_execute_command', command=command,
+                          arguments=arguments)
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 16]
```diff
@@ -205,7 +226,8 @@ def format_range(self, doc_uri, range):
         return self._hook('pyls_format_range', doc_uri, range=range)
 
     def highlight(self, doc_uri, position):
-        return flatten(self._hook('pyls_document_highlight', doc_uri, position=position)) or None
+        return flatten(self._hook('pyls_document_highlight', doc_uri,
+                       position=position)) or None
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 17]
```diff
@@ -214,7 +236,9 @@ def hover(self, doc_uri, position):
     def lint(self, doc_uri):
         # Since we're debounced, the document may no longer be open
         if doc_uri in self.workspace.documents:
-            self.workspace.publish_diagnostics(doc_uri, flatten(self._hook('pyls_lint', doc_uri)))
+            self.workspace.publish_diagnostics(doc_uri,
+                                               flatten(self._hook('pyls_lint',
+                                                                  doc_uri)))
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 18]
```diff
@@ -223,7 +247,8 @@ def references(self, doc_uri, position, exclude_declaration):
         ))
 
     def rename(self, doc_uri, position, new_name):
-        return self._hook('pyls_rename', doc_uri, position=position, new_name=new_name)
+        return self._hook('pyls_rename', doc_uri, position=position,
+                          new_name=new_name)
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 19]
```diff
@@ -232,11 +257,13 @@ def m_text_document__did_close(self, textDocument=None, **_kwargs):
         self.workspace.rm_document(textDocument['uri'])
 
     def m_text_document__did_open(self, textDocument=None, **_kwargs):
-        self.workspace.put_document(textDocument['uri'], textDocument['text'], version=textDocument.get('version'))
+        self.workspace.put_document(textDocument['uri'], textDocument['text'],
+                                    version=textDocument.get('version'))
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 20]
```diff
@@ -232,11 +257,13 @@ def m_text_document__did_close(self, textDocument=None, **_kwargs):
...
         self._hook('pyls_document_did_open', textDocument['uri'])
         self.lint(textDocument['uri'])
 
-    def m_text_document__did_change(self, contentChanges=None, textDocument=None, **_kwargs):
+    def m_text_document__did_change(self, contentChanges=None,
+                                    textDocument=None, **_kwargs):
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 21]
```diff
@@ -248,43 +275,54 @@ def m_text_document__did_change(self, contentChanges=None, textDocument=None, **
     def m_text_document__did_save(self, textDocument=None, **_kwargs):
         self.lint(textDocument['uri'])
 
-    def m_text_document__code_action(self, textDocument=None, range=None, context=None, **_kwargs):
+    def m_text_document__code_action(self, textDocument=None, range=None,
+                                     context=None, **_kwargs):
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 22]
```diff
@@ -248,43 +275,54 @@ def m_text_document__did_change(self, contentChanges=None, textDocument=None, **
...
         return self.code_actions(textDocument['uri'], range, context)
 
     def m_text_document__code_lens(self, textDocument=None, **_kwargs):
         return self.code_lens(textDocument['uri'])
 
-    def m_text_document__completion(self, textDocument=None, position=None, **_kwargs):
+    def m_text_document__completion(self, textDocument=None, position=None,
+                                    **_kwargs):
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 23]
```diff
@@ -248,43 +275,54 @@ def m_text_document__did_change(self, contentChanges=None, textDocument=None, **
...
         return self.completions(textDocument['uri'], position)
 
-    def m_text_document__definition(self, textDocument=None, position=None, **_kwargs):
+    def m_text_document__definition(self, textDocument=None, position=None,
+                                    **_kwargs):
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 24]
```diff
@@ -248,43 +275,54 @@ def m_text_document__did_change(self, contentChanges=None, textDocument=None, **
...
         return self.definitions(textDocument['uri'], position)
 
-    def m_text_document__document_highlight(self, textDocument=None, position=None, **_kwargs):
+    def m_text_document__document_highlight(self, textDocument=None,
+                                            position=None, **_kwargs):
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 25]
```diff
@@ -248,43 +275,54 @@ def m_text_document__did_change(self, contentChanges=None, textDocument=None, **
...
         return self.highlight(textDocument['uri'], position)
 
-    def m_text_document__hover(self, textDocument=None, position=None, **_kwargs):
+    def m_text_document__hover(self, textDocument=None, position=None,
+                               **_kwargs):
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 26]
```diff
@@ -248,43 +275,54 @@ def m_text_document__did_change(self, contentChanges=None, textDocument=None, **
...
         return self.hover(textDocument['uri'], position)
 
     def m_text_document__document_symbol(self, textDocument=None, **_kwargs):
         return self.document_symbols(textDocument['uri'])
 
-    def m_text_document__formatting(self, textDocument=None, _options=None, **_kwargs):
+    def m_text_document__formatting(self, textDocument=None, _options=None,
+                                    **_kwargs):
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 27]
```diff
@@ -248,43 +275,54 @@ def m_text_document__did_change(self, contentChanges=None, textDocument=None, **
...
         # For now we're ignoring formatting options.
         return self.format_document(textDocument['uri'])
 
-    def m_text_document__rename(self, textDocument=None, position=None, newName=None, **_kwargs):
+    def m_text_document__rename(self, textDocument=None, position=None,
+                                newName=None, **_kwargs):
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 28]
```diff
@@ -248,43 +275,54 @@ def m_text_document__did_change(self, contentChanges=None, textDocument=None, **
...
         return self.rename(textDocument['uri'], position, newName)
 
-    def m_text_document__range_formatting(self, textDocument=None, range=None, _options=None, **_kwargs):
+    def m_text_document__range_formatting(self, textDocument=None, range=None,
+                                          _options=None, **_kwargs):
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 29]
```diff
@@ -248,43 +275,54 @@ def m_text_document__did_change(self, contentChanges=None, textDocument=None, **
...
         # Again, we'll ignore formatting options for now.
         return self.format_range(textDocument['uri'], range)
 
-    def m_text_document__references(self, textDocument=None, position=None, context=None, **_kwargs):
+    def m_text_document__references(self, textDocument=None, position=None,
+                                    context=None, **_kwargs):
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 30]
```diff
@@ -248,43 +275,54 @@ def m_text_document__did_change(self, contentChanges=None, textDocument=None, **
...
         exclude_declaration = not context['includeDeclaration']
-        return self.references(textDocument['uri'], position, exclude_declaration)
+        return self.references(textDocument['uri'], position,
+                               exclude_declaration)
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/python_ls.py [hunk 31]
```diff
@@ -248,43 +275,54 @@ def m_text_document__did_change(self, contentChanges=None, textDocument=None, **
...
 
-    def m_text_document__signature_help(self, textDocument=None, position=None, **_kwargs):
+    def m_text_document__signature_help(self, textDocument=None, position=None,
+                                        **_kwargs):
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Revert because it's an stylistic change.

File: pyls/workspace.py [hunk 32]
```diff
@@ -15,11 +17,60 @@
...
+        submods = pkgutil.walk_packages(m.__path__, m.__name__ + '.',
+                                        catch_exceptions)
+        for sm in submods:
+            sm_name = sm[1]
+            submodules.append(sm_name)
+    except ImportError:
+        return []
+    except Exception:
+        return [mod]
+    return submodules
```
Comment (ccordoba12) [inline pyls/workspace.py]: Please remove these two functions from here (i.e. `get_preferred_submodules` and `get_submodules`). There's no need to have them here because we can change the list of preloaded (for Jedi) and extension (for Rope) modules from the Spyder side, by using the PyLS config system and our Preferences.

File: pyls/workspace.py [hunk 33]
```diff
@@ -39,7 +90,9 @@ def _rope_project_builder(self, rope_config):
         if self.__rope is None or self.__rope_config != rope_config:
             rope_folder = rope_config.get('ropeFolder')
             self.__rope = Project(self._root_path, ropefolder=rope_folder)
-            self.__rope.prefs.set('extension_modules', rope_config.get('extensionModules', []))
+            # self.__rope.prefs.set('extension_modules',
+            #                       rope_config.get('extensionModules', []))
+            # TODO: check if the modules are faster.
```
Comment (ccordoba12) [inline pyls/workspace.py]: Please revert this change and leave things as they were before.

File: pyls/workspace.py [hunk 34]
```diff
@@ -15,11 +17,60 @@
...
+        return [mod]
+    return submodules
+
+
 class Workspace(object):
 
     M_PUBLISH_DIAGNOSTICS = 'textDocument/publishDiagnostics'
     M_APPLY_EDIT = 'workspace/applyEdit'
     M_SHOW_MESSAGE = 'window/showMessage'
+    PRELOADED_MODULES = get_preferred_submodules()
```
Comment (ccordoba12) [inline pyls/workspace.py]: Please remove this, it's not needed.

File: pyls/workspace.py [hunk 35]
```diff
@@ -68,7 +121,8 @@ def get_document(self, doc_uri):
         return self._docs.get(doc_uri) or self._create_document(doc_uri)
 
     def put_document(self, doc_uri, source, version=None):
-        self._docs[doc_uri] = self._create_document(doc_uri, source=source, version=version)
+        self._docs[doc_uri] = self._create_document(doc_uri, source=source,
+                                                    version=version)
```
Comment (ccordoba12) [inline pyls/workspace.py]: Revert because it's an stylistic change.

File: pyls/workspace.py [hunk 36]
```diff
@@ -81,14 +135,18 @@ def apply_edit(self, edit):
         return self._endpoint.request(self.M_APPLY_EDIT, {'edit': edit})
 
     def publish_diagnostics(self, doc_uri, diagnostics):
-        self._endpoint.notify(self.M_PUBLISH_DIAGNOSTICS, params={'uri': doc_uri, 'diagnostics': diagnostics})
+        self._endpoint.notify(self.M_PUBLISH_DIAGNOSTICS,
+                              params={'uri': doc_uri,
+                                      'diagnostics': diagnostics})
```
Comment (ccordoba12) [inline pyls/workspace.py]: Revert because it's an stylistic change.

File: pyls/workspace.py [hunk 37]
```diff
@@ -81,14 +135,18 @@ def apply_edit(self, edit):
...
 
     def show_message(self, message, msg_type=lsp.MessageType.Info):
-        self._endpoint.notify(self.M_SHOW_MESSAGE, params={'type': msg_type, 'message': message})
+        self._endpoint.notify(self.M_SHOW_MESSAGE, params={'type': msg_type,
+                                                           'message': message})
```
Comment (ccordoba12) [inline pyls/workspace.py]: Revert because it's an stylistic change.

File: pyls/workspace.py [hunk 38]
```diff
@@ -81,14 +135,18 @@ def apply_edit(self, edit):
...
 
     def source_roots(self, document_path):
         """Return the source roots for the given document."""
-        files = _utils.find_parents(self._root_path, document_path, ['setup.py']) or []
+        files = _utils.find_parents(self._root_path, document_path,
+                                    ['setup.py']) or []
```
Comment (ccordoba12) [inline pyls/workspace.py]: Revert because it's an stylistic change.

File: pyls/workspace.py [hunk 39]
```diff
@@ -102,7 +160,8 @@ def _create_document(self, doc_uri, source=None, version=None):
 
 class Document(object):
 
-    def __init__(self, uri, source=None, version=None, local=True, extra_sys_path=None, rope_project_builder=None):
+    def __init__(self, uri, source=None, version=None, local=True,
+                 extra_sys_path=None, rope_project_builder=None):
```
Comment (ccordoba12) [inline pyls/workspace.py]: Revert because it's an stylistic change.

File: pyls/workspace.py [hunk 40]
```diff
@@ -210,7 +269,9 @@ def jedi_script(self, position=None):
         }
         if position:
             kwargs['line'] = position['line'] + 1
-            kwargs['column'] = _utils.clip_column(position['character'], self.lines, position['line'])
+            kwargs['column'] = _utils.clip_column(position['character'],
+                                                  self.lines,
+                                                  position['line'])
```
Comment (ccordoba12) [inline pyls/workspace.py]: Revert because it's an stylistic change.

**Review by ccordoba12 — COMMENTED** [review]
@steff456, I left an initial review for you. Once you revert all stylistic changes you made and try to restore current functionality, I'll take another look at it.

File: pyls/plugins/jedi_completion.py (hunk 1 above)
Comment (steff456) [reply pyls/plugins/jedi_completion.py]: I remove it because spyder is waiting for a list and not for a None. So when the function returned None, the autocompletion in spyder crashed. My solution was to return an empty list so the completion will not be compromised in our side.

File: pyls/plugins/rope_completion.py (hunk 3 above)
Comment (steff456) [reply pyls/plugins/rope_completion.py]: I did this change because they refactor the signature for the pyls completion in jedi. So when the race started, the arguments for jedi were working but the ones for rope wasn't. In this way both completions have the same signature.

File: pyls/plugins/rope_completion.py (hunk 4 above)
Comment (steff456) [reply pyls/plugins/rope_completion.py]: This was the same solution for the problem I stated before of the completion crashing in spyder because we are expecting a list, not a None.

File: pyls/plugins/rope_completion.py (hunk 6 above)
Comment (steff456) [reply pyls/plugins/rope_completion.py]: Same as the above explanation.

File: pyls/python_ls.py (hunk 13 above)
Comment (steff456) [reply pyls/python_ls.py:189]: This change is neccessary when the completions function returns a None, this change can be reverted if we don't return a None on the completions function.

File: pyls/plugins/jedi_completion.py (hunk 1 above)
Comment (ccordoba12) [reply pyls/plugins/jedi_completion.py]: Then that's something we need to handle on Spyder's side, not here.

The idea is that we need to remain as compatible as possible with the PyLS developed by Palantir, so we can keep merging changes their changes here. So we can't break compatibility like this.

File: pyls/plugins/rope_completion.py (hunk 4 above)
Comment (ccordoba12) [reply pyls/plugins/rope_completion.py]: As I explained above, please revert this to maintain compatibility with upstream's PyLS.

File: pyls/plugins/rope_completion.py (hunk 3 above)
Comment (ccordoba12) [reply pyls/plugins/rope_completion.py]: Ok, that's fine then.

Comment (ccordoba12) [reply pyls/plugins/rope_completion.py]: But is it possible for you to leave things like this?

    def pyls_completions(document, position, config, workspace)

That's because both `config` and `workspace` are mandatory (not optional) for Rope's `pyls_completion`.

File: pyls/plugins/rope_completion.py [hunk 41]
```diff
@@ -15,15 +15,17 @@ def pyls_settings():
...
-def pyls_completions(config, workspace, document, position):
+def pyls_completions(document, position, config=None, workspace=None):
+    log.debug('Launching rope ... ')
     # Rope is a bit rubbish at completing module imports, so we'll return None
     word = document.word_at_position({
         # The -1 should really be trying to look at the previous word, but that might be quite expensive
         # So we only skip import completions when the cursor is one space after `import`
-        'line': position['line'], 'character': max(position['character'] - 1, 0),
+        'line': position['line'],
+        'character': max(position['character'] - 1, 0),
```
Comment (ccordoba12) [inline pyls/plugins/rope_completion.py]: Please revert this change because it's stylistic.

File: pyls/python_ls.py [hunk 42]
```diff
@@ -111,11 +112,15 @@ def m_exit(self, **_kwargs):
...
+        return self.config.plugin_manager.subset_hook_caller(hook_name, self.config.disabled_plugins)
+
     def _hook(self, hook_name, doc_uri=None, **kwargs):
         """Calls hook_name and returns a list of results from all registered handlers"""
         doc = self.workspace.get_document(doc_uri) if doc_uri else None
-        hook_handlers = self.config.plugin_manager.subset_hook_caller(hook_name, self.config.disabled_plugins)
-        return hook_handlers(config=self.config, workspace=self.workspace, document=doc, **kwargs)
+        return self._hook_caller(hook_name)(config=self.config,
+                                            workspace=self.workspace,
+                                            document=doc, **kwargs)
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Please revert this change and remove the `_hook_caller` method you added above.

File: pyls/workspace.py [hunk 43]
```diff
@@ -3,6 +3,8 @@
 import logging
 import os
 import re
+import imp
+import pkgutil
```
Comment (ccordoba12) [inline pyls/workspace.py]: Remove these import's because they are not needed anymore.

File: pyls/python_ls.py (hunk 42 above)
Comment (ccordoba12) [reply pyls/python_ls.py]: @steff456, please address this comment as well (perhaps you missed it? ;-)

File: pyls/python_ls.py [hunk 44]
```diff
@@ -111,11 +112,17 @@ def m_exit(self, **_kwargs):
...
+        return self.config.plugin_manager.subset_hook_caller(hook_name, self.config.disabled_plugins)
+
     def _hook(self, hook_name, doc_uri=None, **kwargs):
         """Calls hook_name and returns a list of results from all registered handlers"""
         doc = self.workspace.get_document(doc_uri) if doc_uri else None
         hook_handlers = self.config.plugin_manager.subset_hook_caller(hook_name, self.config.disabled_plugins)
         return hook_handlers(config=self.config, workspace=self.workspace, document=doc, **kwargs)
+        # return self._hook_caller(hook_name)(config=self.config,
+        #                                     workspace=self.workspace,
+        #                                     document=doc, **kwargs)
```
Comment (ccordoba12) [inline pyls/python_ls.py]: You leave this as a comment. Please remove it.

File: pyls/python_ls.py [hunk 45]
```diff
@@ -111,11 +112,17 @@ def m_exit(self, **_kwargs):
         self._jsonrpc_stream_reader.close()
         self._jsonrpc_stream_writer.close()
 
+    def _hook_caller(self, hook_name):
+        return self.config.plugin_manager.subset_hook_caller(hook_name, self.config.disabled_plugins)
```
Comment (ccordoba12) [inline pyls/python_ls.py]: And also remove this method because it's also not necessary.

File: pyls/plugins/jedi_completion.py [hunk 46]
```diff
@@ -6,16 +6,19 @@
...
+    } for d in definitions]
```
Comment (ccordoba12) [inline pyls/plugins/jedi_completion.py]: Revert this change.

File: pyls/plugins/rope_completion.py [hunk 47]
```diff
@@ -15,15 +15,16 @@ def pyls_settings():
...
+    log.debug('Launching rope ... ')
     # Rope is a bit rubbish at completing module imports, so we'll return None
     word = document.word_at_position({
         # The -1 should really be trying to look at the previous word, but that might be quite expensive
         # So we only skip import completions when the cursor is one space after `import`
         'line': position['line'], 'character': max(position['character'] - 1, 0),
     })
     if word == 'import':
-        return None
+        return []
```
Comment (ccordoba12) [inline pyls/plugins/rope_completion.py]: Revert this change

File: pyls/plugins/rope_completion.py [hunk 48]
```diff
@@ -51,8 +52,10 @@ def pyls_completions(config, workspace, document, position):
             'sortText': _sort_text(d)
         })
     definitions = new_definitions
-
-    return definitions or None
+    log.debug('Finish Rope!!!')
+    if not definitions:
+        return []
+    return definitions
```
Comment (ccordoba12) [inline pyls/plugins/rope_completion.py]: Here leave the call to `log.debug` then leave things as before:

    return definitions or None

File: pyls/python_ls.py [hunk 49]
```diff
@@ -183,11 +185,26 @@ def code_lens(self, doc_uri):
...
+                self._hook_caller('pyls_completions'),
+                self._pool,
+                document=self.workspace.get_document(doc_uri) if doc_uri else None,
+                position=position,
+                config=self.config,
+                workspace=self.workspace
+            )
+            return {
+                'isIncomplete': False,
+                'items': completions
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Please  change this line to

    'items': flatten(completions) if completions else None

File: pyls/python_ls.py [hunk 50]
```diff
@@ -183,11 +185,26 @@ def code_lens(self, doc_uri):
...
+                }
+        else:
+            completions = self._hook('pyls_completions', doc_uri, position=position)
+            return {
+                'isIncomplete': False,
+                'items': flatten(completions)
```
Comment (ccordoba12) [inline pyls/python_ls.py]: And the same here

    'items': flatten(completions) if completions else None

**Review by ccordoba12 — COMMENTED** [review]
Last review (I think). This should maintain compatibility with Palantir and fix the error with `flatten` at the same time.

**Comment by ccordoba12** [comment]
@steff456, I'm going to try to activate Circle and Appveyor to run tests here, so I'm going to close/reopen this PR several times, ok?

File: pyls/python_ls.py [hunk 51]
```diff
@@ -183,11 +185,26 @@ def code_lens(self, doc_uri):
...
+                self._pool,
+                document=self.workspace.get_document(doc_uri) if doc_uri else None,
+                position=position,
+                config=self.config,
+                workspace=self.workspace
+            )
+            return {
+                'isIncomplete': False,
+                'items': flatten(completions) if completions else None
+                }
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Let's remove this `return` because we have another `return` which is exactly the same below.

File: pyls/python_ls.py [hunk 52]
```diff
@@ -183,11 +185,26 @@ def code_lens(self, doc_uri):
...
+        else:
+            completions = self._hook('pyls_completions', doc_uri, position=position)
+            return {
+                'isIncomplete': False,
+                'items': flatten(completions) if completions else None
+                }
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Please move this `return` out of the `if/else` block. We only need that block to get the right assignment for  `completions`.

File: pyls/_utils.py [hunk 53]
```diff
@@ -131,3 +131,26 @@ def is_process_alive(pid):
...
+    def _apply(impl):
+        return impl, impl.function(**kwargs)
+
+    # imap unordered gives us an iterator over the items in the order they finish.
+    # We have to be careful to set chunksize to 1 to ensure hooks each get their own thread.
+    # Unfortunately, there's no way to interrupt these threads, so we just have to leave them be.
+    first_impl, result = next(pool.imap_unordered(_apply, impls, chunksize=1))
+    log.debug("Hook from plugin %s returned: %s", first_impl.plugin_name,
+              result)
+    return result
```
Comment (ccordoba12) [inline pyls/_utils.py]: Please add an empty new line below this one.

File: pyls/python_ls.py [hunk 54]
```diff
@@ -183,11 +185,26 @@ def code_lens(self, doc_uri):
...
     def completions(self, doc_uri, position):
-        completions = self._hook('pyls_completions', doc_uri, position=position)
-        return {
-            'isIncomplete': False,
-            'items': flatten(completions)
-        }
+        rope_enabled = self.config.settings()['plugins']['rope_completion']['enabled']
+        if rope_enabled:
+            completions = _utils.race_hooks(
+                self._hook_caller('pyls_completions'),
```
Comment (ccordoba12) [inline pyls/python_ls.py]: This needs to be `self._hook` now, not `self._hook_caller`. Please change it.

File: pyls/python_ls.py [hunk 55]
```diff
@@ -155,6 +156,7 @@ def m_initialize(self, processId=None, rootUri=None, rootPath=None, initializati
         self.workspace = Workspace(rootUri, self._endpoint)
         self.config = config.Config(rootUri, initializationOptions or {}, processId)
         self._dispatchers = self._hook('pyls_dispatchers')
+        self._pool = multiprocessing.Pool(PLUGGY_RACE_POOL_SIZE)
```
Comment (ccordoba12) [inline pyls/python_ls.py:160]: Please add this line

    self._pool = None

to the `__init__` method (below `self.config = None`) to avoid the pylint warning.

File: pyls/plugins/jedi_completion.py [hunk 56]
```diff
@@ -6,16 +6,19 @@
 
 
 @hookimpl
-def pyls_completions(document, position):
+def pyls_completions(document, position, config=None, workspace=None):
```
Comment (ccordoba12) [inline pyls/plugins/jedi_completion.py:13]: Please add a comment at the top of this file, like this

    # Copyright 2017 Palantir Technologies, Inc.

    # Don't complain about unused arguments
    # pylint: disable=W0613

This will avoid the Pylint warning about unused arguments.

**Review by ccordoba12 — COMMENTED** [review]
@steff456, my suggested changes should help you to pass the style issues reported in Circle.

File: pyls/plugins/jedi_completion.py [hunk 57]
```diff
@@ -1,21 +1,27 @@
 # Copyright 2017 Palantir Technologies, Inc.
+
+# Don't complain about unused arguments
+# pylint: disable=W0613
```
Comment (ccordoba12) [inline pyls/plugins/jedi_completion.py:4]: Please add a blank line below this comment.

**Review by ccordoba12 — COMMENTED** [review]
Last comment and this should be ready! Yay!!

**Comment by ccordoba12** [comment]
Thanks a lot @steff456!
